### ISSUE TITLE
takePickture issue

### ISSUE THREAD

**Issue by aliwaqassi** [cordova-plugin-camera-preview/cordova-plugin-camera-preview#221]
once picture is captured then the preview show the original image, when i again click on take picture app is crashed, fix this please
also switch camera is not working it stuck the camera when click on it

**Comment by gorghoa** [comment]
Hi, I also experiment crash when leveraging the `takePicture` method.

I’ve tested only on android 5.1

Here is the options I use:

```js
document.addEventListener("deviceready", () => {
            CameraPreview.startCamera({x: 0, y: 0,
                width: width,
                height: height,
                camera: "rear",
                toBack: true,
                tapPhoto: false,
                previewDrag: false
            }, () => {
                 this.filming = true;

            }, (e:any) => console.error(e));
        });
```

I can see the preview, it’s pretty neat by the way, congrats ;)

later: 

```js
CameraPreview.takePicture((res) => {
                    console.log('#OK');
                }, (e) => {
                     console.error(e);
                });
```

crash with “Unfortunately, xxx has stopped”, the app is closed

**Comment by aliwaqassi** [comment]
what is the solution?

**Comment by gorghoa** [comment]
@aliwaqassi, I’ve no solution yet, I was just adding some more infos about the bug we seem to have in common.

**Comment by aliwaqassi** [comment]
hmmm ok , wait for owner reply

**Comment by gorghoa** [comment]
@aliwaqassi, removing the plugin and reinstalling it with the latest master seems to have fix my problem.

**Comment by aliwaqassi** [comment]
@gorghoa i installed 1 hour ago

**Comment by gorghoa** [comment]
I had a second problem though, maybe it will helps:

When I ask for `takePicture` once and then ask a second `takePicture` before the first one finalized, the app crashed.

I’ve worked around with a lock:

```js
let lock = false;

takeAPicture = () => {
 if (!lock) {
      lock = true;
      CameraPreview.takePicture(options, () => {
          lock = false;
      }, () => {
          lock = false;
      });
};
```

**Comment by gorghoa** [comment]
@westonganger, I experiment something weird, I don’t know if it’s a bug or a feature :), when i take a picture, the preview freeze with the capture displayed, I have to restart the camera (`cameraStop` then `cameraStart`) for another pic, normal?

**Comment by westonganger** [comment]
Yeah I noticed that too. I would like that to be fixed for sure.

@anneb Does removing the stopCamera / startCamera in the android code cause this freezing after takePicture? I thought I saw some comments where this wasn't the case?

**Comment by anneb** [comment]
For Android, some commits ago we removed calls mCamera.stopPreview() and mCamera.startPreview() from takePicture(), because they were causing stability problems for some users. These calls were taken from other source code where it was claimed to prevent stability problems.

I am not sure if this takePicture issue is about the Android version or the iOS version or both?

**Comment by westonganger** [comment]
I think this is Android being referred to. In your Android testing does the preview freeze after takePicture?

**Comment by bhambanip** [comment]
Yes, it's freezing after taking pictures, is there any solution to it?

**Comment by anneb** [comment]
see PR #225 to fix the preview freeze

**Comment by anneb** [comment]
On Android, the crash mentioned in this issue can be reproduced by taking high resolution portrait pictures.
A Galaxy S5 app crashes when trying to rotate a 5312x2988 resolution picture to portrait orientation on calling:
matrix.postRotate(angle);
return Bitmap.createBitmap(source, 0, 0, source.getWidth(), source.getHeight(), matrix, true);

Maybe http://stackoverflow.com/questions/17169339/rotating-bitmap-causes-outofmemoryexception may help fix the problem, but the accepted answer is not complete and my Java async memory management knowledge fails me here completely.

quick hack: pass in a smaller width/height when calling CameraPreview.takePicture, for instance:
CameraPreview.takePicture({width: 2048, height: 2048}, function (base64Data) {...});

**Comment by westonganger** [comment]
PR #225 has been merged.

**Comment by aliwaqassi** [comment]
thanks All it works.

**Comment by anneb** [comment]
added PR #228 for Android error handling if the picture is too large (better than app crash).

todo: on devices with high resolution cameras, the maximum resolution pictures are often too large to be properly handled in memory. 

Ideas?
* set default maximum resolution to 2 Megapixel?
* automatically downsample picture until it fits in memory (= resource intensive)?
* require options and require options.width and options.height to be specified?
* use file storage for large picture operations (requires additional file access permission)?
* other solution?

**Comment by gorghoa** [comment]
@anneb , as a user, setting the default maximum resolution to 2Megapixel seems reasonable, and far better than a crash ;)

It may be combined with your other proposal: requiring width and height.

To the least, these two options are very explicit for the developer using your plugin.

Thanks for your work, the takePicture freeze is gone ;)

**Comment by anneb** [comment]
add PR #230 for limiting default picture resolution to max 2 Megapixel
Some commonly expected maximum default resolutions:
ratio 4:3 => 1440x1080 or 1600x1200
ratio 16:9 => 1920x1080

**Comment by anneb** [comment]
The behaviour of the Android and iOS versions of the plugin are different for the the following situation:
A camera preview of 100 x 200 pixels where the user requests a takePicture of 10 x 10 pixels

To compare with the current behavior of the iOS version of the plugin, I checked the iOS version on an iPhone 5s and found the following:
* by default the iOS version creates photos of 3264x2448 pixels (4:3). The photo resolution or photo size does not change if the preview size or preview aspect ratio is changed.  If the preview is not 4:3 (or 3:4), then the resulting photo contains image portions that were not visible in the preview.
* if width and height options are passed in, the image is resized so that one of the dimensions matches the requested width or height. The other dimension is calculated such that the resulting picture aspect ratio is always 4:3 (or 3:4)

Conclusion: Android and iOS versions are currently (and have long been) different:
iOS: 4:3 (or 3:4) pictures only
iOS: options width and height are interpreted like maxWidth and maxHeight, the images is scaled so that one of the dimensions exactly matches the requested maxWidth or maxHeight and the other dimension assures the aspect ratio remains 4:3 (or 3:4)
Android: both 4:3 (or 3:4), 16:9 (or 9:16) or other supported aspect ratios available, wichever is closest to the preview aspect ratio
Android: options width and height are interpreted like maxWidth and maxHeight, the resulting picture is the largest supported resolution image below options 'width' and 'height' that is closest to the preview aspect ratio, the image is not scaled to match at least one of the requested dimensions.

Android and iOS behave identically when:
* preview aspect ratio is 4:3 or 3:4
* takePicture width and height options match aspect ratio 4:3 or 3:4
* the takePicture width and height match a supported picture resolution on Android

For all other situations, I believe it is necessary to define:
How exactly should the plugin behave for a camera preview of 400 x 200 pixels where the user requests a takePicture of 50 x 60 pixels?

**Comment by westonganger** [comment]
Wow this is really complicated. I feel like this needs to be standardized & documented so we get an expected output on both devices.

I have now made you a collaborator as you've been working so hard on this plugin. As your contributing to this keep in mind the goal is to provide the most similarity between the two platforms. Also provide documentation in the readme whenever possible. Thanks!

**Comment by westonganger** [comment]
The out of memory error should be caught also. See this commit for the implementation: https://github.com/suddrey/cordova-plugin-camera-preview/commit/1c679afddc3be9d64f2e3eb760c1ced813e26df1

**Comment by luckylooke** [comment]
I have similar issue that takePicture() crashes app, I found out that timeouting takePicture from startCamera execution has solved the issue. I have used timeout for 3000ms.

**Comment by tomcatmwi** [comment]
On my Motorola XT1034 the takePicture() function works properly, unless the app is sent into the background and kept alive with BackgroundMode. If it's in the background, the app crashes. I am trying to develop a timelapse app, but it doesn't seem to be possible without keeping the device constantly awake.

**Comment by Shmarkus** [comment]
for me, the problem is memory usage. Taking full reso pictures (android) and from the profiler, I see that memory crawls to 250 and then the app is stopped.
